Secondary source statistics (v2)

bbockelm opened the pull request:
This breaks out the FJR information (and corresponding HTCondor statistics) for the secondary source versus all sources.

The intent is to provide insight for amount of input data for PileUp-vs-Primary in various DIGI-RECO workflows.

This is a cleaned-up version of #20107; I believe I got all the review comments addressed from that PR.

bbockelm committed:
Fix C++11 items from clang-tidy.

bbockelm committed:
Switch labels to an enumerated, more explicit context.

This turns the labels from std::strings to an enum class.  Changes
the API to allow the context to be seen more explicitly.

bbockelm committed:
Add back const'ness.

bbockelm committed:
Cache token inside TStorageFactoryFile.

Additionally, include the remaining operations in the list of all
metrics.

bbockelm committed:
Add statistics aggregation for the secondary source.

This breaks out the FJR information (and corresponding HTCondor
statistics) for the secondary source versus all sources.

The intent is to provide insight for amount of input data for
PileUp-vs-Primary in various DIGI-RECO workflows.

Conflicts:
	Utilities/StorageFactory/src/StorageAccount.cc

cmsbuild commented:
A new Pull Request was created by  @bbockelm (Brian Bockelman) for master.

It involves the following packages:

FWCore/Services
IOPool/Input
IOPool/TFileAdaptor
Utilities/StorageFactory


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20149/81

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20149/81/git-diff.patch

In future, you can run `scram build code-checks` to apply code checks

bbockelm committed:
Apply suggested changes from `clang-tidy`.

cmsbuild commented:
Pull request #20149 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pr-code-checks/PR-20149/88

Dr15Jones commented on Utilities/StorageFactory/src/StorageAccount.cc:
```diff
@@ -78,35 +140,48 @@ const std::string& StorageAccount::nameForToken( StorageClassToken iToken) {
   std::ostringstream os;
   if (banner)
     os << "stats: class/operation/attempts/successes/amount/time-total/time-min/time-max\n";
   for (auto i = s_nameToToken.begin (); i != s_nameToToken.end(); ++i) {
```
For future reference, a 'range based for loop' would probably make this logic cleaner

Dr15Jones commented on Utilities/StorageFactory/src/StorageAccount.cc:
```diff
@@ -78,35 +140,48 @@ const std::string& StorageAccount::nameForToken( StorageClassToken iToken) {
   for (auto i = s_nameToToken.begin (); i != s_nameToToken.end(); ++i) {
     auto const& opStats = m_stats[i->second];
-    for (auto j = opStats.begin (); j != opStats.end (); ++j, first = false)
+    for (auto j = opStats.begin (); j != opStats.end (); ++j, first = false) {
```
A range based for loop would help here as well

Dr15Jones commented on Utilities/StorageFactory/src/StorageAccount.cc:
```diff
@@ -78,35 +140,48 @@ const std::string& StorageAccount::nameForToken( StorageClassToken iToken) {
+
   int const oneM = 1000 * 1000;
   int const oneMeg = 1024 * 1024;
   for (auto i = s_nameToToken.begin (); i != s_nameToToken.end(); ++i) {
```
A range based for would be nice here as well.

Dr15Jones commented on Utilities/StorageFactory/src/StorageAccount.cc:
```diff
@@ -50,23 +86,49 @@ static std::string d2str(double d) {
+    protocols.insert(item.first.second);
+  }
+  for (auto const &protocol : protocols) {
+    auto token = tokenForStorageClassName(protocol, OpenLabel::None);
```
So is `OpenLabel::None` just meant to be used to denote an aggregation across all contexts? If so, maybe a more descriptive name would help clarify the logic?

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4797f6155967f71ace9985661f7042dceb8ee9b2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22324/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7d6f190a9a22b134851a160272dc06c46c1a5063
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22324/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22324/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22324/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2718193
 - DQMHistoTests: Total failures: 259
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2717745
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

Dr15Jones commented:
@bbockelm waiting on a reply to my comments.

bbockelm committed:
Rename `None` to `Aggregate` in enum.

The `None` member of the storage statistics enum was utilized to
hold the aggregate statistics of all access types; this commit renames
it to `Aggregate` in order to make the point of the member clearer.

bbockelm committed:
Convert traditional for loops to ranged for-loops.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
Pull request #20149 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20149/512

bbockelm commented:
@Dr15Jones - I think I hit all the review items in the last two commits.  Since the branch was originally forked from an IB, I also rebased off the newer CMSSW_9_3_0_pre5 -- hence all the recent commits listed in the github UI.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d21a767fc3864edbd2dbe4c06977e2ff40166a2b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22749/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b9e49920ac6fb7c8b41a37df4151eae9de6776ea
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22749/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22749/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20149/22749/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2656498
 - DQMHistoTests: Total failures: 209
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2656100
 - DQMHistoTests: Total skipped: 189
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 14 edm output root files, 26 DQM output files

Dr15Jones commented on Utilities/StorageFactory/interface/StorageAccount.h:
```diff
@@ -36,6 +38,7 @@ class StorageAccount {
     writeViaCache,
     writev
```
If you add a `nOperations` item at the end, you can use that value in the `allOperations` array declaration. That solves the problem you were trying to avoid in your comments.

davidlange6 commented:
hi @bbockelm @Dr15Jones - any plans to move forward with this one?

Dr15Jones commented:
There is still an open question to Brian.

davidlange6 commented:
hi @bbockelm - any plans to update this PR?

Dr15Jones commented:
@bbockelm ping

fabiocos commented:
@bbockelm @Dr15Jones is this PR still of interest? Any plan to move forward with it?

bbockelm commented:
While I'd really like to get this one across the finish line, I really have no time leftover at the end of the day.  Going to close now to help clean up the list of open PRs; hopefully some day I can revisit!
